kayenta-prometheus/src/main/java/com/netflix/kayenta/prometheus/metrics/PrometheusMetricDescriptorsCache.java

```diff
@@ -0,0 +1,131 @@
+                            metricDescriptor ->
+                                    metricDescriptor.getName().toLowerCase().contains(lowerCaseFilter))
+                    .map(metricDescriptor -> metricDescriptor.getMap())
+                    .collect(Collectors.toList());
+        }
+    }
+
+    @Scheduled(fixedDelayString = "#{@prometheusConfigurationProperties.metadataCachingIntervalMS}")
+    public void updateMetricDescriptorsCache() {
+        Set<AccountCredentials> accountCredentialsSet =
+                accountCredentialsRepository.getAllOf(AccountCredentials.Type.METRICS_STORE);
+
+        Map<String, List<PrometheusMetricDescriptor>> updatedCache = accountCredentialsSet
+                .stream()
+                .filter(credentials -> credentials instanceof PrometheusNamedAccountCredentials)
+                .map(credentials -> (PrometheusNamedAccountCredentials) credentials)
+                .map(this::listMetricDescriptors)
+                .filter(this::isSuccessful)
+                .filter(this::hasData)
+                .collect(Collectors.toMap(
+                        AccountResponse::getMetricsAccountName,
+                        this::toPrometheusMetricDescriptors));
+
+        this.cache = updatedCache;
+    }
+
+    private List<PrometheusMetricDescriptor> toPrometheusMetricDescriptors(AccountResponse accountResponse) {
+        List<PrometheusMetricDescriptor> descriptors = accountResponse.getResponse().getData().stream()
+                .map(PrometheusMetricDescriptor::new)
+                .collect(Collectors.toList());
+        log.debug(
+                "Updated cache with {} metric descriptors via account {}.",
+                descriptors.size(),
+                accountResponse.getMetricsAccountName());
+        return descriptors;
+    }
+
+    private AccountResponse listMetricDescriptors(PrometheusNamedAccountCredentials credentials) {
+        PrometheusRemoteService prometheusRemoteService = credentials.getPrometheusRemoteService();
+        PrometheusMetricDescriptorsResponse remoteResponse = prometheusRemoteService.listMetricDescriptors();
+        return new AccountResponse(credentials.getName(), remoteResponse);
+    }
+
+    private boolean hasData(AccountResponse accountResponse) {
+        boolean empty = CollectionUtils.isEmpty(accountResponse.getResponse().getData());
+        if (empty) {
+            log.debug(
+                    "While updating cache, found no metric descriptors via account {}.", accountResponse.getMetricsAccountName());
+        }
+        return !empty;
+    }
+
+    private boolean isSuccessful(AccountResponse accountResponse) {
+        PrometheusMetricDescriptorsResponse response = accountResponse.getResponse();
+        boolean success = response != null && response.getStatus().equals("success");
+        if (!success) {
+            log.debug("While updating cache, found no metric descriptors via account {}.", accountResponse.getMetricsAccountName());
```

Comment:
this was the original message in this code, but I suppose it should be something different, because the request actually failed. please let me know if I need to change it or leave it as is